client/consensus/common/src/level_monitor.rs

```diff
@@ -0,0 +1,269 @@
+						// a descendant leaf fresher (or equally fresh) than the current candidate.
+						if leaf_idx < candidate_freshest_leaf_idx &&
+							route.common_block().hash != best_hash &&
+							route.enacted().iter().all(|entry| entry.hash != best_hash)
+						{
+							// We have a new candidate
+							candidate_freshest_leaf_idx = leaf_idx;
+							candidate_freshest_route = Some(route);
+						}
+						break
+					},
+					Err(err) => {
+						log::warn!(
+							target: "parachain",
+							"Unable getting route from {} to {}: {}",
+							blk_hash, leaf_hash, err,
+						);
+					},
+					_ => (),
+				};
+			}
+			if candidate_freshest_leaf_idx == 0 {
+				// We can't find a candidate with an older leaf.
+				break
+			}
+		}
+
+		let candidate_freshest_route = match candidate_freshest_route {
+			Some(route) => route,
+			None => return,
+		};
+
+		// We have a candidate, proceed removing the block and all its descendants.
+
+		let candidate_hash = candidate_freshest_route.common_block().hash;
+
+		// Takes care of route removal. Starts from the leaf and stops as soon as an error is
+		// encountered. In this case an error is interpreted as the block being not a leaf
+		// and it will be removed while removing another route from the same block but to a
+		// different leaf.
+		let mut remove_route = |route: TreeRoute<Block>| {
+			std::iter::once(route.common_block()).chain(route.enacted()).rev().all(|elem| {
+				if self.backend.remove_leaf_block(&elem.hash).is_err() {
+					return false
+				}
+				log::debug!(target: "parachain", "Removing block {}", elem.hash);
+				self.levels
+					.get_mut(&elem.number)
+					.and_then(|level| level.remove(&elem.hash).then_some(()));
+				self.freshness.remove(&elem.hash);
+				true
+			});
+		};
+
+		remove_route(candidate_freshest_route);
+
+		let to_skip = leaves.len() - candidate_freshest_leaf_idx;
+		leaves.iter().rev().skip(to_skip).for_each(|leaf_hash| {
+			match sp_blockchain::tree_route(blockchain, candidate_hash, *leaf_hash) {
+				Ok(route) if route.retracted().is_empty() => {
+					remove_route(route);
+				},
+				Err(err) => {
+					log::warn!(
+						target: "parachain",
+						"Unable getting route from {} to {}: {}",
+						candidate_hash, leaf_hash, err,
+					);
+				},
+				_ => (),
+			};
+		});
+	}
+
+	/// Add a new imported block information to the monitor.
+	pub fn update(&mut self, number: NumberFor<Block>, hash: Block::Hash) {
+		self.freshness.insert(hash, self.import_counter);
+		self.levels.entry(number).or_default().insert(hash);
+		self.import_counter += 1;
+
+		if self.import_counter % CLEANUP_THRESHOLD == 0 {
+			// Do cleanup once in a while, we are allowed to have some obsolete information.
+			let finalized_num = self.backend.blockchain().info().finalized_number;
+			let delta: u64 =
+				finalized_num.saturating_sub(self.lowest_level).unique_saturated_into();
+
+			for i in 0..delta {
+				let number = finalized_num + i.unique_saturated_into();
```

Comment:
`Range` implements `Iterator` only if its `Item` associated type implements `Step`:
https://doc.rust-lang.org/std/ops/struct.Range.html#impl-Iterator

But unfortunately our generic `Number` is not bound by `Step`.

Nor can be manually implemented because it requires the unstable feature `step_trait`:
https://github.com/rust-lang/rust/issues/42168

Step is only directly implemented by Rust std/core for primitive numeric types (https://doc.rust-lang.org/core/iter/trait.Step.html#implementors).

If there is a workaround or if I miss something please correct me.